subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java

```diff
@@ -0,0 +1,138 @@
+
+import static org.gradle.api.internal.initialization.transform.BaseInstrumentingArtifactTransform.InstrumentArtifactTransformParameters;
+import static org.gradle.internal.classpath.TransformedClassPath.INSTRUMENTED_JAR_DIR_NAME;
+import static org.gradle.internal.classpath.TransformedClassPath.INSTRUMENTED_MARKER_FILE_NAME;
+import static org.gradle.internal.classpath.TransformedClassPath.ORIGINAL_JAR_DIR_NAME;
+
+/**
+ * Base artifact transform that instruments plugins with Gradle instrumentation, e.g. for configuration cache detection or property upgrades.
+ */
+@DisableCachingByDefault(because = "Instrumented jars are too big to cache")
+public abstract class BaseInstrumentingArtifactTransform implements TransformAction<InstrumentArtifactTransformParameters> {
+
+    public interface InstrumentArtifactTransformParameters extends TransformParameters {
+        @Input
+        Property<Boolean> getAgentSupported();
+
+        @Input
+        Property<Integer> getMaxSupportedJavaVersion();
+
+        @Input
+        @Optional
+        Property<String> getUpgradedPropertiesHash();
+    }
+
+    @Inject
+    public abstract ObjectFactory getObjects();
+
+    @PathSensitive(PathSensitivity.NAME_ONLY)
+    @InputArtifact
+    public abstract Provider<FileSystemLocation> getInput();
+
+    @Override
+    public void transform(TransformOutputs outputs) {
+        File input = getInput().get().getAsFile();
+        if (!input.exists()) {
+            // Don't instrument files that don't exist, these could be files added to classpath via files()
+            return;
+        }
+
+        // A marker file that indicates that the result is instrumented jar,
+        // this is important so TransformedClassPath can correctly filter instrumented jars.
+        createNewFile(outputs.file(INSTRUMENTED_MARKER_FILE_NAME));
+
+        // Instrument jars
+        InstrumentationServices instrumentationServices = getObjects().newInstance(InstrumentationServices.class);
+        File outputFile = outputs.file(INSTRUMENTED_JAR_DIR_NAME + "/" + input.getName());
+        ClasspathElementTransformFactory transformFactory = instrumentationServices.getTransformFactory(getParameters().getAgentSupported().get());
+        ClasspathElementTransform transform = transformFactory.createTransformer(input, new InstrumentingClassTransform(), InstrumentingTypeRegistry.EMPTY);
+        transform.transform(outputFile);
+
+        // Copy original jars after in case they are not in global cache
+        if (instrumentationServices.getGlobalCacheLocations().isInsideGlobalCache(input.getAbsolutePath())) {
+            // The global caches are additive only, so we can use it directly since it shouldn't be deleted or changed during the build.
+            outputs.file(getInput());
+        } else {
+            File copyOfOriginalFile = outputs.file(ORIGINAL_JAR_DIR_NAME + "/" + input.getName());
+            GFileUtils.copyFile(input, copyOfOriginalFile);
```

Comment:
❓ I asked this before: do we need to copy the original file here? We have also the option to pass `input` directly to `outputs.file()` and then save the time needed to copy the file.